kong/tools/queue.lua

```diff
@@ -0,0 +1,408 @@
+-- system.  If two plugin instances use the same queue name, they will share one
+-- queue and their queue related configuration must match.
+--
+-- `handler` is the function which consumes entries.
+--
+-- Entries are dequeued in batches.  The maximum size of each batch of entries
+-- that is passed to the `handler` function can be configured using the
+-- `batch_max_size` parameter.
+--
+-- The handler function can either return true if the entries
+-- were successfully processed, or false and an error message to indicate that
+-- processing has failed.  If processing has failed, the queue library will
+-- automatically retry.
+--
+-- If the `batch_max_size` parameter is larger than 1, processing of the
+-- queue will only start after `max_delay` milliseconds have elapsed if
+-- less than `batch_max_size` entries are waiting on the queue.  If `batch_max_size`
+-- is 1, the queue will be processed immediately.
+--
+-- Usage:
+--
+--   local Queue = require "kong.tools.queue"
+--
+--   local handler = function(conf, entries)
+--     -- must return true if ok, or false + error otherwise
+--     return true
+--   end
+--
+--   local handler_conf = {...}  -- configuration for queue handler
+--   local queue_conf =          -- configuration for the queue itself (defaults shown unless noted)
+--     {
+--       name = "example",       -- name of the queue (required)
+--       batch_max_size = 10,    -- maximum number of entries in one batch (default 1)
+--       max_delay = 1,          -- maximum number of seconds after first entry before a batch is sent
+--       capacity = 10,          -- maximum number of entries on the queue (default 10000)
+--       string_capacity = 100,  -- maximum number of bytes on the queue (default nil)
+--       max_retry_time = 60,    -- maximum number of seconds before a failed batch is dropped
+--       max_retry_delay = 60,   -- maximum delay between send attempts
+--     }
+--
+--   Queue.enqueue(queue_conf, handler, handler_conf, "Some value")
+--   Queue.enqueue(queue_conf, handler, handler_conf, "Another value")
+--
+-- Given the example above,
+--
+-- * If the two `enqueue()` invocations are done within `max_delay` seconds, they will be passed to the
+--   handler function together in one batch.  The maximum number of entries in one batch is defined
+--   by the `batch_max_size` parameter.
+-- * The `capacity` parameter defines how many entries can be waiting on the queue for transmission.  If
+--   it is exceeded, the oldest entries on the queue will be discarded when new entries are queued.  Error
+--   messages describing how many entries were lost will be logged.
+-- * The `string_capacity` parameter, if set, indicates that no more than that number of bytes can be
+--   waiting on a queue for transmission.  If it is set, only string entries must be queued and an error
+--   message will be logged if an attempt is made to queue a non-string entry.
```

Comment:
Why is an error message logged rather than returned back to the caller of enqueue()?

Reply:
There is really nothing a caller can do facing this configuration error - Likewise for other errors that can be encountered in `enqueue()`.  I can propagate these errors up and duplicate the log statement in the plugins if that is what you prefer.